src/physics/base/ParticleTrackView.hh

```diff
@@ -0,0 +1,87 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file ParticleTrackView.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "ParticleStatePointers.hh"
+#include "ParticleParamsPointers.hh"
+#include "base/Macros.hh"
+#include "base/Types.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Read/write view to the physical properties of a single particle track.
+ *
+ * These functions should be used in each physics Process.
+ */
+class ParticleTrackView
+{
+  public:
+    //@{
+    //! Type aliases
+    using Initializer_t = ParticleTrackState;
+    //@}
+
+  public:
+    // Construct from "dynamic" state and "static" particle definitions
+    inline CELER_FUNCTION
+    ParticleTrackView(const ParticleParamsPointers& params,
+                      const ParticleStatePointers&  states,
+                      ThreadId                      id);
+
+    // Initialize the particle
+    inline CELER_FUNCTION ParticleTrackView&
+                          operator=(const Initializer_t& other);
+
+    // >>> DYNAMIC PROPERTIES
+
+    // Unique particle type identifier
+    inline CELER_FUNCTION ParticleDefId def_id() const;
+
+    // Kinetic energy [MeV]
+    inline CELER_FUNCTION real_type energy() const;
+
+    // >>> STATIC PROPERTIES
+
+    // Rest mass [MeV / c^2]
+    inline CELER_FUNCTION real_type mass() const;
+
+    // Charge [elemental charge e+]
+    inline CELER_FUNCTION real_type charge() const;
+
+    // Decay constant [1/s]
+    inline CELER_FUNCTION real_type decay_constant() const;
+
+    // >>> DERIVED PROPERTIES
+
+    // Speed [1/c]
+    inline CELER_FUNCTION real_type speed() const;
```

Comment:
All accessors look the same externally but have very different computing complexity (eg. speed vs energy).   From a functional point of view the difference is an implementation detail (eventually the speed could be cached) but since we are focusing on performance it might be convenient to be able to quickly find/notice the expensive calls. [The downside would be a required refactoring of the using code when upgrading the performance of a given routine]

Reply:
Well, in addition to performance we're looking for extensibility. I know that geant4/v/x all cache some of these attributes, but I thought we could start out by generating all these on the fly, and experiment with cacheing some or all of them as needed.

We might encourage client code to "cache" the function results as local variables rather than reevaluating them, but that's kind of a good practice anyway:
```c++
const real_type speed = track.speed();
real_type distance = speed * distance;
// something else that uses speed
```

Reply:
> We might encourage client code to "cache" t

That is necessary but not sufficient as some of those value will be used multiple time during a step by different entities (for example by many model during their own GPIL calculations and/or doit).

Especially since those code line will be inlined, it will be "challenging" to later on find out which ones are slowing us down (i.e. the profiler wont be able to aggregate them).

Reply:
Closing after our zoom discussion -- it should be easy enough to cache/replace these one by one.